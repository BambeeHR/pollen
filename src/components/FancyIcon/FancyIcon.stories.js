import { number } from '@storybook/addon-knobs';
import FancyIcon, { Icons } from './FancyIcon.vue';

export default {
  title: 'FancyIcon',
  component: FancyIcon,
  parameters: {
    componentSubtitle: 'Inlined svg icons',
  },
};

export const Gallery = () => ({
  components: { FancyIcon },
  props: {
    size: {
      default: number('Size', 72),
    },
  },
  data() {
    return {
      icons: Object.values(Icons),
    };
  },
  template: `
    <div class="flex flex-wrap">
      <div v-for="icon in icons" :key="icon" class="text-center m-3">
        <div :style="{ fontSize: size + 'px' }"><FancyIcon :icon="icon" /></div>
        <div class="font-body-small mt-2">{{ icon }}</div>
      </div>
    </div>
  `,
});
