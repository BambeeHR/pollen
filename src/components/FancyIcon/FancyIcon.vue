<template functional>
  <component
    :is="$options.getComponent(props.icon)"
    :class="[data.class, data.staticClass]"
    class="fancy-icon"
    v-bind="$options.getAttributes({ data, parent })"
    v-on="listeners"
  />
</template>

<script>
import camelCase from 'lodash/camelCase';
import snakeCase from 'lodash/snakeCase';

import clipboard from '../../assets/icons/fancy/clipboard.svg';
import signTablet from '../../assets/icons/fancy/sign-tablet.svg';

import getAttributes from '../../util/getAttributes';

const IconComponents = Object.freeze({
  clipboard,
  signTablet,
});

export const Icons = Object.freeze(
  Object.keys(IconComponents).reduce(
    (obj, val) => ({
      ...obj,
      [snakeCase(val).toUpperCase()]: val,
    }),
    {}
  )
);

export const isValidIcon = (value) =>
  Object.values(Icons).includes(camelCase(value));

const getComponent = (iconName) => IconComponents[iconName];

/**
 * `import { FancyIcon } from '@bambee/pollen';`
 *
 * An SVG icon component pulling from Bambee's predefined multi-color icon set.
 * These are sized using contextual font size, but unlike BaseIcons, their
 * colors will not change.
 */
export default {
  props: {
    /** The name of the icon to use. */
    icon: {
      type: String,
      required: true,
      validator: isValidIcon,
    },
  },
  getAttributes,
  getComponent,
};
</script>

<style>
.fancy-icon {
  @apply block leading-none;
  height: 1em;
  width: 1em;
}
</style>
