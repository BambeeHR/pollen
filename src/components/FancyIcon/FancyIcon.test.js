import { mount } from '@vue/test-utils';
import { Gallery } from './FancyIcon.stories';

describe('FancyIcon', () => {
  test('renders gallery correctly', () => {
    const gallery = Gallery();
    const wrapper = mount(gallery);
    expect(wrapper.element).toMatchSnapshot();
  });
});
