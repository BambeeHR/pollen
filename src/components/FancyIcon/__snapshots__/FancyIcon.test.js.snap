// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FancyIcon renders gallery correctly 1`] = `
<div
  class="flex flex-wrap"
>
  <div
    class="text-center m-3"
  >
    <div
      style="font-size: 72px;"
    >
      <svg
        class="fancy-icon"
      />
    </div>
     
    <div
      class="font-body-small mt-2"
    >
      clipboard
    </div>
  </div>
  <div
    class="text-center m-3"
  >
    <div
      style="font-size: 72px;"
    >
      <svg
        class="fancy-icon"
      />
    </div>
     
    <div
      class="font-body-small mt-2"
    >
      signTablet
    </div>
  </div>
</div>
`;
